hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenAirdropHandler.java

```diff
@@ -0,0 +1,567 @@
+        requireNonNull(txn);
+        final var op = txn.tokenAirdropOrThrow();
+        validator.airdropsPureChecks(op);
+    }
+
+    @Override
+    public void handle(@NonNull final HandleContext context) throws HandleException {
+        requireNonNull(context);
+        final var txn = context.body();
+        final var op = txn.tokenAirdropOrThrow();
+        final var pendingStore = context.storeFactory().writableStore(WritableAirdropStore.class);
+        final var accountStore = context.storeFactory().writableStore(WritableAccountStore.class);
+        var recordBuilder = context.recordBuilders().getOrCreate(TokenAirdropRecordBuilder.class);
+        List<TokenTransferList> tokenTransferList = new ArrayList<>();
+
+        // charge custom fees in advance
+        var convertedOp = CryptoTransferTransactionBody.newBuilder()
+                .tokenTransfers(op.tokenTransfers())
+                .build();
+        assessCustomFee(context, convertedOp);
+
+        for (final var xfers : op.tokenTransfers()) {
+            final var tokenId = xfers.tokenOrThrow();
+
+            boolean shouldExecuteCryptoTransfer = false;
+            var transferListBuilder = TokenTransferList.newBuilder().token(tokenId);
+
+            // process fungible token transfers if any
+            if (!xfers.transfers().isEmpty()) {
+                // 1. separate transfers in to two lists
+                // - one list for executing the transfer and one list for adding to pending state
+                var fungibleLists = separateFungibleTransfers(context, tokenId, xfers.transfers());
+                var senderOptionalAmount = xfers.transfers().stream()
+                        .filter(item -> item.amount() < 0)
+                        .findFirst();
+                var senderAccount = accountStore.getForModify(
+                        senderOptionalAmount.orElseThrow().accountIDOrThrow());
+                validateTrue(senderAccount != null, INVALID_ACCOUNT_ID);
+
+                // 2. create and save pending airdrops in to state
+                fungibleLists.pendingFungibleAmounts().forEach(accountAmount -> {
+                    var pendingId = createFungibleTokenPendingAirdropId(
+                            tokenId, senderOptionalAmount.orElseThrow().accountID(), accountAmount.accountID());
+                    var pendingValue = PendingAirdropValue.newBuilder()
+                            .amount(accountAmount.amount())
+                            .build();
+                    AccountPendingAirdrop newAccountPendingAirdrop = getNewAccountPendingAirdropAndUpdateStores(
+                            senderAccount, pendingId, pendingValue, accountStore, pendingStore);
+                    var record = createPendingAirdropRecord(pendingId, newAccountPendingAirdrop.pendingAirdropValue());
+                    pendingStore.put(pendingId, newAccountPendingAirdrop);
```

Comment:
⚠️ With NFT airdrops we validated the airdrop didn't already exist; is it intentional to overwrite an existing pending airdrop with the same key for fungible tokens?

Reply:
Well, we don't overwrite it, as the pendingStore.put method it adds to the PendingValue for fungible token, so if we have pending airdrop with 10 tokens in store and we airdrop another 20, at the end we keep 30 in the store.